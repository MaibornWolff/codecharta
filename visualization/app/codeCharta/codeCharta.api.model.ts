import { AttributeTypes, AttributeTypeValue, CodeMapNode, Edge, MarkedPackage, RecursivePartial, Settings } from "./codeCharta.model"

export interface ExportCCFile {
	projectName: string
	apiVersion: string
	nodes: CodeMapNode[]
	attributeTypes?: AttributeTypes | OldAttributeTypes
	edges?: Edge[]
	markedPackages?: MarkedPackage[]
	blacklist?: ExportBlacklistItem[]
}

export interface ExportBlacklistItem {
	path: string
	type: ExportBlacklistType
}

export enum ExportBlacklistType {
	hide = "hide",
	exclude = "exclude"
}

export enum APIVersions {
	ZERO_POINT_ONE = "0.1",
	ONE_POINT_ZERO = "1.0",
	ONE_POINT_ONE = "1.1"
}

export interface ExportScenario {
	name: string
	settings: RecursivePartial<Settings>
}

export interface OldAttributeTypes {
	nodes?: [{ [key: string]: AttributeTypeValue }]
	edges?: [{ [key: string]: AttributeTypeValue }]
}
