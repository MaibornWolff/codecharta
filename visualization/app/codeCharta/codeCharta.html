<div id="hud" class="container">

    <div class="row">

        <div class="col l3 m6 s12 col-backdrop">
            <legend-panel-component></legend-panel-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <revision-chooser-component></revision-chooser-component>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <file-chooser-panel-directive></file-chooser-panel-directive>
        </div>

        <div class="col l3 m6 s12 col-backdrop">
            <settings-panel-component></settings-panel-component>
        </div>

    </div>

    <div class="row bottom-row">

        <div class="col l3 m6 s12 col-backdrop">
            <detail-panel-component></detail-panel-component>
        </div>

    </div>

    <div id="titlebox">
        <a class="mwlink" href="https://www.maibornwolff.de" target="_blank"><img src="./logo.png" alt="MaibornWolff"></a>
        <h1 id="title"><span>CodeCharta {{$ctrl.pkg.version}}</span></h1>
    </div>

</div>

<code-map-component></code-map-component>