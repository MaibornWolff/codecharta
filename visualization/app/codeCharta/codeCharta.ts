"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent, CodeChartaController} from "./codeCharta.component";

import angular from "angular";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
);


