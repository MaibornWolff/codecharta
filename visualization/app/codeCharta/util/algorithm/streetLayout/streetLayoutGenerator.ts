import { CodeMapNode, Node, State, NodeMetricData, BlacklistType, LayoutAlgorithm } from "../../../codeCharta.model"
import BoundingBox from "./boundingBox"
import VerticalStreet from "./verticalStreet"
import HorizontalStreet from "./horizontalStreet"
import House from "./house"
import TreeMap from "./treeMap"
import { Vector2 } from "three"
import { StreetOrientation } from "./street"
import { getMapResolutionScaleFactor, isPathBlacklisted, isLeaf } from "../../codeMapHelper"
import { StreetViewHelper } from "./streetViewHelper"
import SquarifiedTreeMap from "./squarifiedTreeMap"

const MARGIN_SCALING_FACTOR = 0.02
const HEIGHT_SCALING_FACTOR = 0.1
export class StreetLayoutGenerator {
	static createStreetLayoutNodes(map: CodeMapNode, state: State, metricData: NodeMetricData[], isDeltaState: boolean): Node[] {
		const mapSizeResolutionScaling = getMapResolutionScaleFactor(state.files)
		const maxHeight =
			(metricData.find(x => x.name === state.dynamicSettings.heightMetric).maxValue * mapSizeResolutionScaling) /
			HEIGHT_SCALING_FACTOR
		const heightScale = (state.treeMap.mapSize * 2) / maxHeight

		const metricName = state.dynamicSettings.areaMetric
		const mergedMap = StreetViewHelper.mergeDirectories(map, metricName)
		const maxTreeMapFiles = state.appSettings.maxTreeMapFiles
		const childBoxes = this.createBoxes(mergedMap, metricName, state, StreetOrientation.Vertical, 0, maxTreeMapFiles)
		const rootStreet = new HorizontalStreet(mergedMap, childBoxes, 0)
		rootStreet.calculateDimension(metricName)
		const margin = state.dynamicSettings.margin * MARGIN_SCALING_FACTOR
		const layoutNodes = rootStreet.layout(margin, new Vector2(0, 0))

		return layoutNodes.map(streetLayoutNode => {
			return StreetViewHelper.buildNodeFrom(streetLayoutNode as CodeMapNode, heightScale, maxHeight, state, isDeltaState)
		})
	}

	private static createBoxes(
		node: CodeMapNode,
		metricName: string,
		state: State,
		orientation: StreetOrientation,
		depth: number,
		maxTreeMapFiles: number
	): BoundingBox[] {
		const children: BoundingBox[] = []
		const areaMetric = state.dynamicSettings.areaMetric
		for (let child of node.children) {
			if (isPathBlacklisted(child.path, state.fileSettings.blacklist, BlacklistType.exclude)) {
				continue
			}
			if (isLeaf(child)) {
				children.push(new House(child))
				continue
			}

			const layoutAlgorithm = state.appSettings.layoutAlgorithm
			const fileDescendants = StreetLayoutGenerator.countFileDescendants(child)
			if (layoutAlgorithm === LayoutAlgorithm.TreeMapStreet && fileDescendants <= maxTreeMapFiles) {
				const treeMap = StreetLayoutGenerator.createTreeMap(child)
				children.push(treeMap)
			} else {
				child = StreetViewHelper.mergeDirectories(child, areaMetric)
				const streetChildren = StreetLayoutGenerator.createBoxes(
					child,
					metricName,
					state,
					1 - orientation,
					depth + 1,
					maxTreeMapFiles
				)
				const street = StreetLayoutGenerator.createStreet(child, orientation, streetChildren, depth)
				children.push(street)
			}
		}
		return children
	}

	private static createStreet(node: CodeMapNode, orientation: StreetOrientation, children: BoundingBox[], depth: number) {
		return orientation === StreetOrientation.Horizontal
			? new HorizontalStreet(node, children, depth)
			: new VerticalStreet(node, children, depth)
	}

	private static createTreeMap(node: CodeMapNode): TreeMap {
		return new SquarifiedTreeMap(node)
	}

	private static countFileDescendants(folderNode: CodeMapNode): number {
		let totalFileNodes = 0
		for (const child of folderNode.children) {
			totalFileNodes += isLeaf(child) ? 1 : StreetLayoutGenerator.countFileDescendants(child)
		}
		return totalFileNodes
	}
}
