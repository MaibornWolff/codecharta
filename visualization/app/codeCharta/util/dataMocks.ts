import { CCFile, CodeMapNode, Edge, MetricData, Node, Settings } from "../codeCharta.model"
import { CodeMapBuilding } from "../ui/codeMap/rendering/codeMapBuilding"
import * as THREE from "three"

export const VALID_NODE: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	children: [
		{
			name: "big leaf",
			type: "File",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			children: [
				{
					name: "small leaf",
					type: "File",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				}
			]
		}
	]
}

export const VALID_NODE_WITH_PATH: CodeMapNode = {
	name: "root",
	attributes: {},
	type: "Folder",
	path: "/root",
	children: [
		{
			name: "big leaf",
			type: "File",
			path: "/root/big leaf",
			attributes: { RLOC: 100, Functions: 10, MCC: 1 },
			link: "http://www.google.de"
		},
		{
			name: "Parent Leaf",
			type: "Folder",
			attributes: {},
			path: "/root/Parent Leaf",
			children: [
				{
					name: "small leaf",
					type: "File",
					path: "/root/Parent Leaf/small leaf",
					attributes: { RLOC: 30, Functions: 100, MCC: 100 }
				},
				{
					name: "other small leaf",
					type: "File",
					path: "/root/Parent Leaf/other small leaf",
					attributes: { RLOC: 70, Functions: 1000, MCC: 10 }
				},
				{
					name: "empty folder",
					type: "Folder",
					path: "/root/Parent Leaf/empty folder",
					attributes: {},
					children: []
				}
			]
		}
	]
}

export const VALID_EDGES: Edge[] = [
	{
		fromNodeName: "/root/big leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	},
	{
		fromNodeName: "/root/sample1 only leaf",
		toNodeName: "/root/Parent Leaf/small leaf",
		attributes: {
			pairingRate: 89,
			avgCommits: 34
		}
	}
]

export const VALID_EDGE: Edge = {
	fromNodeName: "/root/big leaf",
	toNodeName: "/root/Parent Leaf/small leaf",
	attributes: {
		pairingRate: 89,
		avgCommits: 34
	}
}

export const TEST_FILE_CONTENT = {
	fileName: "noFileName",
	projectName: "Sample Map",
	apiVersion: "1.1",
	nodes: [VALID_NODE]
}

export const TEST_FILE_DATA: CCFile = {
	fileMeta: {
		fileName: "file",
		projectName: "Sample Map",
		apiVersion: "1.1"
	},
	map: VALID_NODE,
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_WITH_PATHS: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		path: "/root",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				path: "/root/big leaf",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				path: "/root/Parent Leaf",
				children: [
					{
						name: "small leaf",
						type: "File",
						path: "/root/Parent Leaf/small leaf",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						path: "/root/Parent Leaf/other small leaf",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					},
					{
						name: "empty folder",
						type: "Folder",
						path: "/root/Parent Leaf/empty folder",
						attributes: {},
						children: []
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_A: CCFile = {
	fileMeta: {
		fileName: "fileA",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000, mcc: 10 }
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_DELTA_MAP_B: CCFile = {
	fileMeta: {
		fileName: "fileB",
		projectName: "Sample Project",
		apiVersion: "1.1"
	},
	map: {
		name: "root",
		type: "Folder",
		attributes: {},
		children: [
			{
				name: "big leaf",
				type: "File",
				attributes: { rloc: 20, functions: 10, mcc: 1 },
				link: "http://www.google.de"
			},
			{
				name: "additional leaf",
				type: "File",
				attributes: { rloc: 10, functions: 11, mcc: 5 },
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				type: "Folder",
				attributes: {},
				children: [
					{
						name: "small leaf",
						type: "File",
						attributes: { rloc: 30, functions: 100, mcc: 100, more: 20 }
					},
					{
						name: "other small leaf",
						type: "File",
						attributes: { rloc: 70, functions: 1000 }
					},
					{
						name: "big leaf",
						type: "File",
						attributes: { rloc: 20, functions: 10, mcc: 1 },
						link: "http://www.google.de"
					}
				]
			}
		]
	},
	settings: {
		fileSettings: {
			attributeTypes: {},
			blacklist: [],
			edges: VALID_EDGES,
			markedPackages: []
		}
	}
}

export const TEST_FILE_DATA_DOWNLOADED = {
	apiVersion: "1.1",
	attributeTypes: {},
	blacklist: [],
	edges: [
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/big leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		},
		{
			attributes: { avgCommits: 34, pairingRate: 89 },
			fromNodeName: "/root/sample1 only leaf",
			toNodeName: "/root/Parent Leaf/small leaf"
		}
	],
	fileName: "file_2018-12-14_9-39.cc.json",
	nodes: [
		{
			attributes: {},
			children: [
				{
					attributes: { Functions: 10, MCC: 1, RLOC: 100 },
					link: "http://www.google.de",
					name: "big leaf",
					type: "File"
				},
				{
					attributes: {},
					children: [
						{
							attributes: { Functions: 100, MCC: 100, RLOC: 30 },
							name: "small leaf",
							type: "File"
						},
						{
							attributes: { Functions: 1000, MCC: 10, RLOC: 70 },
							name: "other small leaf",
							type: "File"
						}
					],
					name: "Parent Leaf",
					type: "Folder"
				}
			],
			name: "root",
			type: "Folder"
		}
	],
	projectName: "Sample Map"
}

export const SETTINGS: Settings = {
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	dynamicSettings: {
		areaMetric: "rloc",
		heightMetric: "mcc",
		colorMetric: "mcc",
		focusedNodePath: "/root",
		searchedNodePaths: [],
		searchPattern: "",
		margin: 48,
		neutralColorRange: { flipped: false, from: 19, to: 67 }
	},
	appSettings: {
		amountOfTopLabels: 31,
		scaling: new THREE.Vector3(1, 1.8, 1),
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		maximizeDetailPanel: false,
		invertHeight: true,
		dynamicMargin: true,
		isWhiteBackground: false,
		whiteColorBuildings: true,
		mapColors: {
			positive: "#69AE40",
			neutral: "#ddcc00",
			negative: "#820E0E",
			selected: "#EB8319",
			defaultC: "#89ACB4",
			positiveDelta: "#69FF40",
			negativeDelta: "#ff0E0E",
			base: "#666666",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			angularGreen: "#00BFA5",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"]
		}
	},
	treeMapSettings: { mapSize: 500 }
}

export const DEFAULT_SETTINGS: Settings = {
	appSettings: {
		amountOfTopLabels: 1,
		camera: new THREE.Vector3(0, 300, 1000),
		deltaColorFlipped: false,
		dynamicMargin: true,
		enableEdgeArrows: true,
		hideFlatBuildings: true,
		invertHeight: false,
		isWhiteBackground: false,
		mapColors: {
			angularGreen: "#00BFA5",
			base: "#666666",
			defaultC: "#89ACB4",
			flat: "#AAAAAA",
			lightGrey: "#DDDDDD",
			markingColors: ["#FF1D8E", "#1d8eff", "#1DFFFF", "#8eff1d", "#8e1dff", "#FFFF1D"],
			negative: "#820E0E",
			negativeDelta: "#ff0E0E",
			neutral: "#ddcc00",
			positive: "#69AE40",
			positiveDelta: "#69FF40",
			selected: "#EB8319"
		},
		maximizeDetailPanel: false,
		scaling: new THREE.Vector3(1, 1, 1),
		whiteColorBuildings: false
	},
	dynamicSettings: {
		areaMetric: null,
		colorMetric: null,
		focusedNodePath: "",
		heightMetric: null,
		margin: null,
		neutralColorRange: {
			flipped: false,
			from: null,
			to: null
		},
		searchPattern: "",
		searchedNodePaths: []
	},
	fileSettings: { attributeTypes: {}, blacklist: [], edges: [], markedPackages: [] },
	treeMapSettings: { mapSize: 500 }
}

export const TEST_NODE_ROOT: Node = {
	name: "root",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 10,
	visible: true,
	path: "root",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0x000000",
	flat: false
}

export const TEST_NODE_LEAF: Node = {
	name: "root/big leaf",
	width: 1,
	height: 2,
	length: 3,
	depth: 4,
	x0: 5,
	z0: 6,
	y0: 7,
	isLeaf: true,
	deltas: { a: 1, b: 2 },
	attributes: { a: 20, b: 15 },
	children: [],
	parent: undefined,
	heightDelta: 20,
	visible: true,
	path: "root/big leaf",
	origin: "root",
	link: "NO_LINK",
	markingColor: "0xFFFFFF",
	flat: false
}

export const CODE_MAP_BUILDING: CodeMapBuilding = new CodeMapBuilding(
	1,
	new THREE.Box3(),
	TEST_NODE_ROOT,
	DEFAULT_SETTINGS.appSettings.mapColors.neutral
)

export const METRIC_DATA: MetricData[] = [
	{ name: "mcc", maxValue: 1, availableInVisibleMaps: true },
	{ name: "rloc", maxValue: 2, availableInVisibleMaps: true }
]
