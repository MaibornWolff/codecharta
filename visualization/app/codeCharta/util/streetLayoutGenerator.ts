import { CodeMapNode, MetricData, Node, State, BlacklistType, TreeMapAlgorithm, LayoutAlgorithm } from "../codeCharta.model"
import BoundingBox from "./algorithm/streetLayout/boundingBox"
import VerticalStreet from "./algorithm/streetLayout/verticalStreet"
import HorizontalStreet from "./algorithm/streetLayout/horizontalStreet"
import House from "./algorithm/streetLayout/house"
import Rectangle from "./algorithm/rectangle"
import Point from "./algorithm/point"
import { CodeMapHelper } from "./codeMapHelper"
import TreeMap from "./algorithm/treeMap/treeMap"
import SliceDiceTreeMap from "./algorithm/treeMap/sliceDiceTreeMap"
import SquarifiedTreeMap from "./algorithm/treeMap/squarifiedTreeMap"
import StripTreeMap from "./algorithm/treeMap/stripTreeMap"
import { StreetOrientation } from "./algorithm/streetLayout/street"
import { StreetLayoutHelper } from "./streetLayoutHelper"
import { LayoutHelper, LayoutNode } from "./layoutHelper"

export interface StreetLayoutValuedCodeMapNode {
	data: CodeMapNode
	value: number
	rect: Rectangle
	zOffset: number
}

export class StreetLayoutGenerator {
	private static MARGIN_SCALING_FACTOR = 0.02

	public static createStreetLayoutNodes(map: CodeMapNode, state: State, metricData: MetricData[]): Node[] {
		const isDeltaState = state.files.isDeltaState()
		const metricName = state.dynamicSettings.areaMetric
		const mergedMap = LayoutHelper.mergeDirectories(map, metricName)
		const maxTreeMapFiles = state.appSettings.maxTreeMapFiles
		const childBoxes = this.createBoxes(mergedMap, metricName, state, StreetOrientation.Vertical, 0, maxTreeMapFiles)
		const rootStreet = new HorizontalStreet(mergedMap, childBoxes, 0)
		rootStreet.calculateDimension(metricName)
		const margin = state.dynamicSettings.margin * StreetLayoutGenerator.MARGIN_SCALING_FACTOR
		const layoutNodes: StreetLayoutValuedCodeMapNode[] = rootStreet.layout(new Point(0, 0), margin)
		const maxHeight = metricData.find(x => x.name == state.dynamicSettings.heightMetric).maxValue
		const heightScale = 1 //TODO: apply correct height after scaling layout size
		return layoutNodes.map(streetLayoutNode => {
			return LayoutHelper.buildNodeFrom(streetLayoutNode as LayoutNode, heightScale, maxHeight, state, isDeltaState)
		})
	}

	private static createBoxes(
		node: CodeMapNode,
		metricName: string,
		state: State,
		orientation: StreetOrientation,
		depth: number,
		maxTreeMapFiles: number
	): BoundingBox[] {
		const children: BoundingBox[] = []
		const areaMetric = state.dynamicSettings.areaMetric
		for (let child of node.children) {
			if (CodeMapHelper.isBlacklisted(child, state.fileSettings.blacklist, BlacklistType.exclude)) {
				continue
			}
			if (LayoutHelper.isNodeLeaf(child)) {
				children.push(new House(child))
			} else {
				const layoutAlgorithm = state.appSettings.layoutAlgorithm
				const fileDescendants = StreetLayoutHelper.countFileDescendants(child)
				if (layoutAlgorithm === LayoutAlgorithm.TMStreet && fileDescendants <= maxTreeMapFiles) {
					const treeMap = StreetLayoutGenerator.createTreeMap(child, TreeMapAlgorithm.Squarified)
					children.push(treeMap)
				} else {
					child = LayoutHelper.mergeDirectories(child, areaMetric)
					const streetChildren = StreetLayoutGenerator.createBoxes(
						child,
						metricName,
						state,
						1 - orientation,
						depth + 1,
						maxTreeMapFiles
					)
					const street = StreetLayoutGenerator.createStreet(child, orientation, streetChildren, depth)
					children.push(street)
				}
			}
		}
		return children
	}

	private static createStreet(node: CodeMapNode, orientation: StreetOrientation, children: BoundingBox[], depth: number) {
		if (orientation === StreetOrientation.Horizontal) {
			return new HorizontalStreet(node, children, depth)
		} else {
			return new VerticalStreet(node, children, depth)
		}
	}

	private static createTreeMap(node: CodeMapNode, treeMapAlgorithm: TreeMapAlgorithm): TreeMap {
		switch (treeMapAlgorithm) {
			case TreeMapAlgorithm.SliceAndDice:
				return new SliceDiceTreeMap(node)
			case TreeMapAlgorithm.Squarified:
				return new SquarifiedTreeMap(node)
			case TreeMapAlgorithm.Strip:
				return new StripTreeMap(node)
			default:
				throw new Error("TreeMap Algorithm not specified.")
		}
	}
}
