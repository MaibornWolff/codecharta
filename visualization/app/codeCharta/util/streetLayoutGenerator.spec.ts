import { CodeMapNode, MetricData, Node, State } from "../codeCharta.model"
import { TEST_FILE_WITH_PATHS, STATE, VALID_NODE_WITH_PATH, METRIC_DATA } from "./dataMocks"
import { StreetLayoutGenerator } from "./streetLayoutGenerator"
import _ from "lodash"

describe("StreetLayoutGenerator", () => {
	let map: CodeMapNode
	let state: State
	let metricData: MetricData[]
	let codeMapNode: CodeMapNode

	beforeEach(() => {
		restartSystem()
	})

	function restartSystem() {
		map = _.cloneDeep(TEST_FILE_WITH_PATHS.map)
		state = _.cloneDeep(STATE)
		codeMapNode = _.cloneDeep(VALID_NODE_WITH_PATH)
		metricData = _.cloneDeep(METRIC_DATA)
	}

	describe("create street layout nodes", () => {
		it("only root node", () => {
			map.children = []

			let nodes: Node[] = StreetLayoutGenerator.createStreetLayoutNodes(map, state, metricData)

			expect(nodes).toMatchSnapshot()
		})

		it("root node with two direct children", () => {
			map.children[1].children = []

			let nodes: Node[] = StreetLayoutGenerator.createStreetLayoutNodes(map, state, metricData)

			expect(nodes).toMatchSnapshot()
		})

		it("root node with two direct children and some grand children", () => {
			let nodes: Node[] = StreetLayoutGenerator.createStreetLayoutNodes(map, state, metricData)

			expect(nodes).toMatchSnapshot()
		})
	})
})
