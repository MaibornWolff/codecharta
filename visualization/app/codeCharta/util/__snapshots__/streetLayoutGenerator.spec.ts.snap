// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StreetLayoutGenerator create street layout nodes only root node 1`] = `
Array [
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -250,
      "y": 2,
      "z": -245.5,
    },
    "isLeaf": true,
    "length": 10,
    "link": undefined,
    "markingColor": null,
    "name": "root",
    "outgoingEdgePoint": Vector3 {
      "x": -250,
      "y": 2,
      "z": -240.5,
    },
    "path": "/root",
    "visible": undefined,
    "width": 0,
    "x0": 0,
    "y0": 2,
    "z0": 0,
  },
]
`;

exports[`StreetLayoutGenerator create street layout nodes root node with two direct children 1`] = `
Array [
  Object {
    "attributes": Object {
      "functions": 10,
      "mcc": 1,
      "rloc": 100,
    },
    "color": "#DDDDDD",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -245,
      "y": 2,
      "z": -245.25,
    },
    "isLeaf": true,
    "length": 9,
    "link": "http://www.google.de",
    "markingColor": null,
    "name": "big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -245,
      "y": 2,
      "z": -240.75,
    },
    "path": "/root/big leaf",
    "visible": undefined,
    "width": 9,
    "x0": 0.5,
    "y0": 2.5,
    "z0": 0,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -245,
      "y": 2,
      "z": -235.5,
    },
    "isLeaf": false,
    "length": 10,
    "link": undefined,
    "markingColor": null,
    "name": "root",
    "outgoingEdgePoint": Vector3 {
      "x": -245,
      "y": 2,
      "z": -230.5,
    },
    "path": "/root",
    "visible": undefined,
    "width": 10,
    "x0": 0,
    "y0": 12,
    "z0": 0,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -250,
      "y": 2,
      "z": -227.75,
    },
    "isLeaf": true,
    "length": -1,
    "link": undefined,
    "markingColor": null,
    "name": "Parent Leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -250,
      "y": 2,
      "z": -228.25,
    },
    "path": "/root/Parent Leaf",
    "visible": undefined,
    "width": -1,
    "x0": 0.5,
    "y0": 22.5,
    "z0": 0,
  },
]
`;

exports[`StreetLayoutGenerator create street layout nodes root node with two direct children and some grand children 1`] = `
Array [
  Object {
    "attributes": Object {
      "functions": 10,
      "mcc": 1,
      "rloc": 100,
    },
    "color": "#DDDDDD",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -245,
      "y": 2,
      "z": -241.4061741596076,
    },
    "isLeaf": true,
    "length": 9,
    "link": "http://www.google.de",
    "markingColor": null,
    "name": "big leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -245,
      "y": 2,
      "z": -236.9061741596076,
    },
    "path": "/root/big leaf",
    "visible": undefined,
    "width": 9,
    "x0": 0.5,
    "y0": 6.343825840392416,
    "z0": 0,
  },
  Object {
    "attributes": Object {
      "functions": 1000,
      "mcc": 10,
      "rloc": 70,
    },
    "color": "#DDDDDD",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -233.81669986732962,
      "y": 2,
      "z": -245.65834993366482,
    },
    "isLeaf": true,
    "length": 7.366600265340756,
    "link": undefined,
    "markingColor": null,
    "name": "other small leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -233.81669986732962,
      "y": 2,
      "z": -241.97504980099444,
    },
    "path": "/root/Parent Leaf/other small leaf",
    "visible": undefined,
    "width": 7.366600265340756,
    "x0": 12.500000000000002,
    "y0": 2.5,
    "z0": 0,
  },
  Object {
    "attributes": Object {
      "functions": 100,
      "mcc": 100,
      "rloc": 30,
    },
    "color": "#820E0E",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -232.3720125221851,
      "y": 2,
      "z": -238.01409334089632,
    },
    "isLeaf": true,
    "length": 4.477225575051661,
    "link": undefined,
    "markingColor": null,
    "name": "small leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -232.3720125221851,
      "y": 2,
      "z": -235.7754805533705,
    },
    "path": "/root/Parent Leaf/small leaf",
    "visible": undefined,
    "width": 4.477225575051661,
    "x0": 15.389374690289095,
    "y0": 10.866600265340756,
    "z0": 0,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -224.63339973465924,
      "y": 2,
      "z": -244.53904353990188,
    },
    "isLeaf": false,
    "length": 13.843825840392416,
    "link": undefined,
    "markingColor": null,
    "name": "Parent Leaf",
    "outgoingEdgePoint": Vector3 {
      "x": -224.63339973465924,
      "y": 2,
      "z": -237.61713061970568,
    },
    "path": "/root/Parent Leaf",
    "visible": undefined,
    "width": 10,
    "x0": 20.366600265340757,
    "y0": 2,
    "z0": 0,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -219.63339973465924,
      "y": 2,
      "z": -233.9061741596076,
    },
    "isLeaf": true,
    "length": -1,
    "link": undefined,
    "markingColor": null,
    "name": "empty folder",
    "outgoingEdgePoint": Vector3 {
      "x": -219.63339973465924,
      "y": 2,
      "z": -234.4061741596076,
    },
    "path": "/root/Parent Leaf/empty folder",
    "visible": undefined,
    "width": -1,
    "x0": 30.866600265340757,
    "y0": 16.343825840392416,
    "z0": 0,
  },
  Object {
    "attributes": Object {},
    "color": "#666666",
    "deltas": undefined,
    "depth": 0,
    "edgeAttributes": undefined,
    "flat": false,
    "height": 2,
    "heightDelta": 0,
    "incomingEdgePoint": Vector3 {
      "x": -242.40834993366482,
      "y": 2,
      "z": -229.1561741596076,
    },
    "isLeaf": false,
    "length": 10,
    "link": undefined,
    "markingColor": null,
    "name": "root",
    "outgoingEdgePoint": Vector3 {
      "x": -227.22504980099444,
      "y": 2,
      "z": -229.1561741596076,
    },
    "path": "/root",
    "visible": undefined,
    "width": 30.366600265340757,
    "x0": 0,
    "y0": 15.843825840392416,
    "z0": 0,
  },
]
`;
